One step forward, an easier interoperability between Rust and Haskell (#23)
input-output-hk/engineering: blog/2023-01-02-hs-bindgen-introduction.md

---
slug: 2022-12-13-ghc-js-backend-merged
title: One step forward, an easier interoperability between Rust and Haskell
authors: [yvan]
tags: [cabal, rust, bindgen]
---

**TL;DR:** This blog post intends to sum up the why and how of [`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal) and [`hs-bindgen`](https://github.com/yvan-sraka/hs-bindgen), if you’re looking for usage walkthroughs and code examples, check out project READMEs on GitHub!

> **N.B.** quoted paragraphs in this article, give straightforward motivation regarding some systems programming basic concepts, feel free to skip them if you know you’re likely to be already comfortable with ;)

## Context

> Here at IOG, we maintain big Haskell codebases that we want to make rely on some Rust dependencies.

Reviewer: ```suggestion
At IOG we maintain large Haskell codebases and we would like to interface them with some libraries written in Rust.
```

Perhaps also a good place to introduce Rust. You're more informed than me but something like this could do:

```
Rust is a relatively new programming language that offers very good performance (comparable to programs written in C) with strong static typing guarantees (similarly to Haskell's). The main difference with Haskell is that there is no garbage collection (the memory must be handled explicitly): that's the reason why programs are usually faster but also more cumbersome to write than with Haskell.
```

> E.g., for cryptographic primitives, which was the first use case that Iñigo (Cardano Lead Cryptography Engineer) brings me: his concerns were to use [`sha3`](https://crates.io/crates/sha3) crate (Rust library) in `cardano-base`, removing the dependency for `cryptonite` (a library written in C).

Reviewer: ```suggestion
One typical use case concerns cryptographic primitives which must be very performant. The first use case that Iñigo NAME (Cardano Lead Cryptography Engineer) brought me consisted in replacing a cryptographic library used by `cardano-base`. Namely, replacing `cryptonite`, a library written in Haskell and C, with [`sha3`](https://crates.io/crates/sha3), a Rust library (or "crate").
```

> **Why FFI (Foreign Function Interface)?**
>

> > Solving the [interoperability](https://docs.rust-embedded.org/book/interoperability/) problem means deciding on what protocol to use to communicate and share data-structure across different runtimes, and (what developers fears most) how do we bundle and distribute such polyglot stack?

Reviewer: ```suggestion
> Solving the [interoperability](https://docs.rust-embedded.org/book/interoperability/) problem means:
> 1. designing a protocol that allows two codes written with different languages and using different runtime systems to communicate
> 2. designing tools and methods to build, to bundle, and to distribute such polyglot code bases (what developers fear most)
```

---
slug: 2023-01-02-hs-bindgen-introduction
title: One step forward, an easier interoperability between Rust and Haskell
authors: [yvan]
tags: [cabal, rust, bindgen]
---

**TL;DR:** This blog post intends to sum up the why and how of [`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal) and [`hs-bindgen`](https://github.com/yvan-sraka/hs-bindgen), if you’re looking for usage walkthroughs and code examples, check out project READMEs on GitHub!

> **N.B.** quoted paragraphs in this article, give straightforward motivation regarding some systems programming basic concepts, feel free to skip them if you know you’re likely to be already comfortable with ;)

## Context

At IOG we maintain large Haskell codebases and we would like to interface them with some libraries written in Rust.

> Rust is a relatively new programming language that offers very good performance (comparable to programs written in C) with strong static typing guarantees (similarly to Haskell's). The main difference with Haskell is that there is no garbage collection (the memory must be handled explicitly): that's the reason why programs are usually faster but also more cumbersome to write than with Haskell.

Author: ```suggestion
Rust is a systems programming language known for its strong static typing guarantees, which make it similar to Haskell (even if a bit less expressive). However, unlike Haskell, Rust does not have a GC (Garbage Collector) and use a compile-time memory management strategy. This mechanism is encoded through its type system with the concept of _“ownership”_ and _“lifetime”_ of values, which can complicate the writing of programs but unlock smaller runtime costs footprint. Rust is becoming increasingly popular in the systems programming and embedded systems domains, and is also used in areas such as cryptography, where performance is critical.
```

FFI is a feature that's already offered by both `rustc` (Rust compiler) and `ghc` (Haskell compiler). It allows calling a Rust function from Haskell code (and vice versa). Both programming languages define an `extern` keyword that allow users to declare a function symbol that will only be resolved at [linking](https://doc.rust-lang.org/reference/linkage.html) step. N.B. mangling of the function should also be disabled, in Rust it requires decorating functions with [`#[no_mangle]`](https://doc.rust-lang.org/reference/abi.html#the-no_mangle-attribute) attribute.

So, what's lacking in Haskell ecosystem? Let's take a look at what kind of integration other languages offer with Rust:

* From **X** to Rust with X=**C**, [`rust-bindgen`](https://github.com/rust-lang/rust-bindgen)

* From Rust to **X** with X=**C**, [`c-bindgen`](https://github.com/eqrion/cbindgen) and X=**ECMAScript**, [`wasm-bindgen`](https://github.com/rustwasm/wasm-bindgen)

* Both with X=**C++**, [`cxx`](https://github.com/dtolnay/cxx) and X=**Python**, [`PyO3`](https://github.com/PyO3/pyo3)

> This list isn't exhaustive but give you a big hint, all these projects are about bindgen!

Reviewer 2: ~big~, generating bindings (bindgen).

> **Why bindgen (bindings code generation)?**
>
> Let's sum it up by: _"A good FFI is an FFI that you don't write …"_
>
> FFI are like a blind spot in your type system. Writing them manually is both frankly painful and really dangerous, as your compiler will not save you of non-matching interfaces.
>
> Binding generation comes to the rescue by considerably reducing the room for human errors, and as a bonus, it simplifies a lot of maintainer job with a more readable reduced codebase.

## Example

So, let's look at a minimal example. In this case, if we annotate a function like this:

```rust
use hs_bindgen::*;

/// Haskell type signature are auto-magically inferred from Rust function
/// type! This feature could slow down compilation, and be enabled with:
/// `hs-bindgen = { ..., features = [ "full" ] }`

> #[hs_bindgen]

Reviewer: Is there a way to specify that we want to generate a `safe` or `unsafe` foreign import in the generated Haskell code? If not, that would be a nice feature to add :)

Author: Not yet! But I will add it to my feature ROADMAP list!

[2] Speaking of memory management, the strategy implemented here is: freeing the value is the role of the receiver (which has “ownership” of it). This means that, e.g. the type that return a Rust function, will not be [`Drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html) by Rust but rather should be [`free`](https://hackage.haskell.org/package/base/docs/Foreign-Marshal-Alloc.html) on Haskell side!

## DevX

[`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal) is a CLI tool that helps you, in one simple command, turning a Rust crate into a Haskell Cabal library!

I was heavily inspired by the developer experience that offers [`wasm-pack`](https://rustwasm.github.io/docs/wasm-pack/) or [`maturin`](https://github.com/PyO3/maturin): launched in any Rust project folder, it helps the user to interactively tweak their `Cargo.toml` to make it successfully build a package compatible with NPM / PyPI and even handle the publishing of it.

What `cargo-cabal` actually does is:

* Ask user to add `crate-type = ["staticlib"]` (or `"cdylib"`, dynamic libraries require an extra `build.rs` that is generated by `cargo-cabal`) to their `Cargo.toml` ;

> * Generate a custom `X.cabal` linking `rustc` output as `extra-librairies`, and either a ([`naersk`](https://github.com/nix-community/naersk) and [`haskell.nix`](https://github.com/input-output-hk/haskell.nix) based) `flake.nix` or a `Setup.lhs` build customization (to work around this [issue](https://github.com/haskell/cabal/issues/2641)).

Reviewer 2: Does it need to be a Setup.lhs? Are we using literate haskell for specific reasons?

Author: No, it could completely be a `Setup.hs`! I just choose to generate literate haskell because I wanted to have a way to differentiate my writings around code of [the Rust build automation commented by default](https://github.com/yvan-sraka/cargo-cabal/blob/main/src/Setup.lhs#L62-L76)!

> 

Author: After discussing in with @angerman, I think I would remove this part, which isn't really accurate (I save it for another post):

> Furthermore, if you need to share large amounts of data between Rust and Haskell, you may want to consider using [Cap’n Proto](https://capnproto.org/): based on memory sharing, prevents (de)serialization cost by offering a common cross-language custom memory mapping. Each language should use the SDK provided to read/write to it, based on RPC communication primitives, to prevent most of the synchronization issues that comes with shared resources approaches. N.B there is already a [Haskell](https://github.com/zenhack/haskell-capnp) and a [Rust](https://github.com/capnproto/capnproto-rust) implementation!

It would be delightful to get as far as having custom preludes in Haskell binding code generated that offers Rust type layout in Haskell. E.g. Rust [slices](https://doc.rust-lang.org/book/ch04-03-slices.html) are not an existing concept in C but could be easily represented as an FFI-safe struct.

Finally, It’s worth mentioning that there are also proposals to improve the interface between Haskell programs requiring Rust libraries, including this Cabal [RFC](https://github.com/haskell/cabal/issues/7906). As a reminder, the implementation proposed here does not provide support for Haskell dependencies in Rust project yet, but there is a previous [unmaintained attempt](https://github.com/mgattozzi/curryrs) of bringing Haskell runtime support in Rust binaries.

> Thanks for reading, feel free to experiment with this Proof of Concept and to provide feedbacks on GitHub :)

Author: I was thinking of giving a list of reference of this form (rather than a bunch of linking hidden in the article corpus), I don't know, what are your thoughts about it?

```suggestion
Thanks for reading, feel free to experiment with this Proof of Concept and to provide feedbacks on GitHub :)

## References

Unordered list of links to go further:

- [`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal): A CLI tool for turning a Rust crate into an Haskell package.
- [`hs-bindgen`](https://github.com/yvan-sraka/hs-bindgen): A tool for generating FFI bindings between Rust and Haskell.
- [`rust-bindgen`](https://github.com/rust-lang/rust-bindgen): A tool for generating FFI bindings between Rust and C.
- [`c-bindgen`](https://github.com/eqrion/cbindgen): A tool for generating FFI bindings between C and Rust.
- [`wasm-bindgen`](https://github.com/rustwasm/wasm-bindgen): A tool for generating FFI bindings between Rust and JavaScript.
- [`cryptonite`](https://github.com/input-output-hk/cardano-crypto/blob/develop/cbits/cryptonite_sha512.h): Cryptographic library used by `cardano-base`.
- [`sha3`](https://crates.io/crates/sha3): A Rust library for sha3 cryptographic primitive.
- [Rust FFI documentation](https://doc.rust-lang.org/nomicon/ffi.html): Foreign Function Interface (FFI) in Unsafe Rust (nomicon) guide.
- [`#[no_mangle]` attribute](https://doc.rust-lang.org/reference/abi.html#the-no_mangle-attribute): An attribute in Rust to prevent symbol mangling.
- [`rustc` linkage](https://doc.rust-lang.org/reference/linkage.html): The process of linking object files together to form an executable program in Rust.
- [`#[repr(interop)]`](https://github.com/rust-lang/rust/pull/105586): Experimental feature gate proposal `interoperable_abi`.
- [_"FFI safety and GC"_ by Fraser Tweedale](https://frasertweedale.github.io/blog-fp/posts/2022-09-23-ffi-safety-and-gc.html): A blog post discussing FFI safety and Garbage Collection in Haskell.
- [Unix Domain Socket](https://en.wikipedia.org/wiki/Unix_domain_socket): A type of inter-process communication socket that allows communication between processes on the same machine.
- [Google Protobuf](https://developers.google.com/protocol-buffers): A protocol for data serialization and communication.
- [Cap'n Proto](https://capnproto.org/): A data interchange format and capability-based RPC system.
- [Haskell Cap'n Proto](https://github.com/zenhack/haskell-capnp): A Haskell implementation of Cap'n Proto.
- [Rust Cap'n Proto](https://github.com/capnproto/capnproto-rust): A Rust implementation of Cap'n Proto.
```

Author: I changed my mind and rather reworked and structure links inside the corpus of the article than dump a big list at the end!

---

> slug: 2023-01-02-hs-bindgen-introduction

Author: ```suggestion
slug: 2023-01-23-hs-bindgen-introduction
```

> slug: 2023-01-23-hs-bindgen-introduction

Author: ```suggestion
slug: 2023-01-24-hs-bindgen-introduction
```

title: One step forward, an easier interoperability between Rust and Haskell
authors: [yvan]
tags: [cabal, rust, bindgen]
---

**TL;DR:** This blog post intends to sum up the why and how of [`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal) and [`hs-bindgen`](https://github.com/yvan-sraka/hs-bindgen), if you’re looking for usage walkthroughs and code examples, check out project READMEs on GitHub!

> **N.B.** quoted paragraphs in this article, give straightforward motivation regarding some systems programming basic concepts, feel free to skip them if you know you’re likely to be already comfortable with ;)

## Context

> At IOG we maintain large Haskell codebases and we would like to interface them with some libraries written in Rust.

Author: ```suggestion
At IOG we maintain large [Haskell](https://www.haskell.org) codebases and we would like to interface them with some libraries written in [Rust](https://www.rust-lang.org).
```

Rust is a systems programming language known for its strong static typing guarantees, which make it similar to Haskell (even if a bit less expressive). However, unlike Haskell, Rust does not have a GC (Garbage Collector) and use a compile-time memory management strategy. This mechanism is encoded through its type system with the concept of _“ownership”_ and _“lifetime”_ of values, which can complicate the writing of programs but unlock smaller runtime costs footprint. Rust is becoming increasingly popular in the systems programming and embedded systems domains, and is also used in areas such as cryptography, where performance is critical.

> One typical use case concerns cryptographic primitives which must be very performant. The first use case that Iñigo Querejeta Azurmendi (Cardano Lead Cryptography Engineer) brought me consisted in replacing a cryptographic library used by `cardano-base`. Namely, replacing `cryptonite`, a library written in Haskell and C, with [`sha3`](https://crates.io/crates/sha3), a Rust library (or "crate").

Author: ```suggestion
One typical use case concerns cryptographic primitives which must be very performant. The first use case that Iñigo Querejeta Azurmendi (Cardano Lead Cryptography Engineer) brought me consisted in replacing a cryptographic library used by [`cardano-base`](https://github.com/input-output-hk/cardano-base). Namely, replacing [`cryptonite`](https://github.com/input-output-hk/cardano-crypto/blob/develop/cbits/cryptonite_sha512.h), a library written in Haskell and C, with [`sha3`](https://crates.io/crates/sha3), a Rust library (or "crate").
```

> **Why FFI (Foreign Function Interface)?**
>
> Solving the [interoperability](https://docs.rust-embedded.org/book/interoperability/) problem means:
> 1. designing a protocol that allows two codes written with different languages and using different runtime systems to communicate
> 2. designing tools and methods to build, to bundle, and to distribute such polyglot code bases (what developers fear most)
>
> Why not just use, e.g., [Google Protobuf](https://developers.google.com/protocol-buffers) over a [Unix Domain Socket](https://en.wikipedia.org/wiki/Unix_domain_socket)? How to make a choice in this landscape of solutions available? First, I will argue that we want something fast, say, by reducing its overhead to the minimal footprint of extra computations. So, we want to avoid use of any OS capabilities, like I/Os (that will cost you a full roundtrip) and the needs (De)Serialization.
>

> > [FFI](https://doc.rust-lang.org/nomicon/ffi.html) looks like the right choice, it doesn’t rely on any syscall, at runtime they just behave as a jump in memory. Unfortunately, it comes at a cost of careful special attention on following call-conventions and handling memory management, we will come back to it later!

Author: ```suggestion
> FFI looks like the right choice, it doesn’t rely on any syscall, at runtime they just behave as a jump in memory. Unfortunately, it comes at a cost of careful special attention on following call-conventions and handling memory management, we will come back to it later!
>
> **To go further:** you can learn more about how to use FFI in Rust by reading the [_The Rustonomicon_](https://doc.rust-lang.org/nomicon/ffi.html) (Unsafe Rust guide) dedicated section, or the dedicated [_Rust FFI Omnibus_](http://jakegoulding.com/rust-ffi-omnibus/) tutorial. The ANSSI (French government security agency) also write about it in [_Secure Rust Guidelines_](https://anssi-fr.github.io/rust-guide/07_ffi.html) guide, and _Rust Embedded_ book have an [Interoperability with C](https://docs.rust-embedded.org/book/interoperability/rust-with-c.html) chapter. On the Haskell side, you way want to take a look at [GHC wiki](https://wiki.haskell.org/GHC/Using_the_FFI) or read the dedicated [_Real-World Haskell_](https://book.realworldhaskell.org/read/interfacing-with-c-the-ffi.html) chapter!
```

-- Why not rather using `{-# LANGUAGE CApiFFI #-}` language extension?
--
-- * Because it's GHC specific and not part of the Haskell standard:
--   https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/ffi.html ;
--
-- * Because the capabilities it gave (by rather works on top of symbols of a C
--   header file) can't work in our case. Maybe we want a future with an
--   {-# LANGUAGE RustApiFFI #-} language extension that would enable us to
--   work on top of a `.rs` source file (or a `.rlib`, but this is unlikely as
--   this format has purposely no public specifications).

{-# OPTIONS_GHC -Wno-unused-imports #-}

module Greetings (greetings) where

import Data.Int
import Data.Word
import Foreign.C.String
import Foreign.C.Types
import Foreign.Ptr

foreign import ccall unsafe "__c_greetings" greetings :: CString -> IO (())
```

In Rust, [`extern`](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html#using-extern-functions-to-call-external-code) is an alias to `extern "C"` that stands for “use the C call convention” rather than `extern "Rust"` that use the Rust one, which is the default implicitly used.

> **Why C ABI (Application Binary Interface)?**
>

> > The Rust [ABI](https://doc.rust-lang.org/reference/items/external-blocks.html#abi) (call-convention and types [memory layout](https://cheats.rs/#memory-layout)) isn’t stable [1]. That means that it’s specified internally but could be broken by any `rustc` minor release, building a software on top of it is by definition a “hack” …

Author: ```suggestion
> The Rust [ABI](https://doc.rust-lang.org/reference/items/external-blocks.html#abi) (call-convention and types [memory layout](https://cheats.rs/#memory-layout)) isn’t stable. That means that it’s specified internally but could be broken by any `rustc` minor release, building a software on top of it is by definition a “hack” …
```

>
> And, if we think it’s worth it, we would have to perform our bindgen against a given `rustc` version (and that would be really laborious to maintain) …
>
> So, do not fear the C ABI … because, at least, it is stable!
>
> [1] https://viruta.org/rust-stable-abi.html

## Implementation

The previous code example highlighted that we rely on two constructs: an attribute procedural macro [`#[hs_bindgen]`](https://github.com/yvan-sraka/hs-bindgen), and (internally) `ReprRust` and `ReprC` traits.

> **Why use a Rust macro?**
>
> Binding code generation could have been achieved using an external tool, e.g., `c-bindgen` parses Rust code (before macro expansion) and deduce the right C function signatures.
>
> But rather we choose to define a custom macro (like `cxx`, `wasm-bindgen` and `PyO3` does), and so we require the user to depend on a custom crate. Why?
>
> We want generated bindings to always match the source code used for generation, so we want to force the user to integrate binding generation in the build process.
>

> > The natural way to do that in Rust, and to have directly access to source code's AST (Abstract Syntax Tree), rather than trying to implement a parser, is to use a macro!

Author: ```suggestion
> The natural way to do that in Rust, and to have directly access to source code's AST (Abstract Syntax Tree), rather than trying to implement a parser, is to use a macro!
>
> **To go further:** `c-bindgen` is well known to be buggy! It's a flaw in its design to have a custom Rust source parser, rather using the framework implement here for `hs-bindgen` would allow us to provide a better `c-bindgen` implementation.
```

Wrapping user types by these traits have several benefits:

* Unsupported types are nicely reported as _“the trait `ReprRust<T>` is not implemented for `U`”_ error (that suggest other types that the trait implement to the user);

* The user can extensively always implement these traits for arbitrary types ;

> * Provided traits implementation for `std` types take care of memory management [2] ;

Author: ```suggestion
* Provided traits implementation for `std` types take care of memory management ;
```

* Traits improve a lot of ergonomics by implicitly and safely casting a given type to an FFI-safe one.

> **What's an FFI-safe type?**
>
> `rustc` will complain if a function prefixed by `extern` keyword use as arguments types that are not FFI-safe. FFI-safe types guarantee that a type has a [specified layout](https://doc.rust-lang.org/reference/type-layout.html) (memory representation) by e.g. having a [`#[repr(C)]`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprc) compiler attribute, for the given C call convention.

> [2] Speaking of memory management, the strategy implemented here is: freeing the value is the role of the receiver (which has “ownership” of it). This means that, e.g. the type that return a Rust function, will not be [`Drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html) by Rust but rather should be [`free`](https://hackage.haskell.org/package/base/docs/Foreign-Marshal-Alloc.html) on Haskell side!

Author: ```suggestion
**To go further:** speaking of memory management, the strategy implemented here is: freeing the value is the role of the receiver (which has “ownership” of it). This means that, e.g. the type that return a Rust function, will not be [`Drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html) by Rust but rather should be [`free`](https://hackage.haskell.org/package/base/docs/Foreign-Marshal-Alloc.html) on Haskell side!
```

## DevX

[`cargo-cabal`](https://github.com/yvan-sraka/cargo-cabal) is a CLI tool that helps you, in one simple command, turning a Rust crate into a Haskell Cabal library!

I was heavily inspired by the developer experience that offers [`wasm-pack`](https://rustwasm.github.io/docs/wasm-pack/) or [`maturin`](https://github.com/PyO3/maturin): launched in any Rust project folder, it helps the user to interactively tweak their `Cargo.toml` to make it successfully build a package compatible with NPM / PyPI and even handle the publishing of it.

What `cargo-cabal` actually does is:

> * Ask user to add `crate-type = ["staticlib"]` (or `"cdylib"`, dynamic libraries require an extra `build.rs` that is generated by `cargo-cabal`) to their `Cargo.toml` ;

Author: ```suggestion
* Ask user to add `crate-type = ["staticlib"]` (or `"cdylib"`, dynamic libraries require an extra [`build.rs`](https://github.com/yvan-sraka/cargo-cabal/blob/main/src/build.rs) that is generated by `cargo-cabal`) to their `Cargo.toml` ;
```

> * Generate a custom `X.cabal` linking `rustc` output as `extra-librairies`, and either a ([`naersk`](https://github.com/nix-community/naersk) and [`haskell.nix`](https://github.com/input-output-hk/haskell.nix) based) `flake.nix` or a `Setup.lhs` build customization (to work around this [issue](https://github.com/haskell/cabal/issues/2641)).

Author: ```suggestion
* Generate a custom `X.cabal` linking `rustc` output as `extra-librairies`, and either a ([`naersk`](https://github.com/nix-community/naersk) and [`haskell.nix`](https://github.com/input-output-hk/haskell.nix) based) `flake.nix` or a [`Setup.lhs`](https://github.com/yvan-sraka/cargo-cabal/blob/main/src/Setup.lhs) build customization (to work around this [issue](https://github.com/haskell/cabal/issues/2641)).
```

> 

Author: ```suggestion

**To go further:** there is no first-class `stack` support yet provided, but we could easily imagine a `cargo-stack` binary that just wrap a `cargo-cabal --stack` CLI option!

```

## Limitations

> What would be the reason to NOT use `cargo-cabal` and `hs-bindgen`? I should mention that `{-# LANGUAGE CApiFFI #-}` language extonsion features, like using C values that are `#define` in a header or `vargs`, are not supported (because they're indeed C specific). E.g. variadic arguments didn't exist in Rust and are usually implemented with [the Builder pattern](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html). Again I want to make obvious that if you're looking for `CApiFFI` features you should embed Rust code in a C library, with e.g. `c-bindgen`, and so not use the tools introduced here!

Author: ```suggestion
What would be the reason to NOT use `cargo-cabal` and `hs-bindgen`? I should mention that `{-# LANGUAGE CApiFFI #-}` language extension features, like using C values that are `#define` in a header or `vargs`, are not supported (because they're indeed C specific). E.g. variadic arguments didn't exist in Rust and are usually implemented with [the Builder pattern](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html). Again I want to make obvious that if you're looking for `CApiFFI` features you should embed Rust code in a C library, with e.g. `c-bindgen`, and so not use the tools introduced here!
```

Furthermore, It’s also nice to give a sneak peek on what others do for comparison: **OCaml** [allows extensions to be written directly in Rust with no C stubs](https://docs.rs/ocaml), this work was supported from the [OCaml Software Foundation](http://ocaml-sf.org/) and you can find [a basic example project here](http://github.com/zshipko/ocaml-rust-starter). It offers [safe OCaml/Rust interoperability](https://github.com/simplestaking/ocaml-interop), meaning utilities to convert ADTs (Algebraic Data Types) and function using them.

It would be delightful to get as far as having custom preludes in Haskell binding code generated that offers Rust type layout in Haskell. E.g. Rust [slices](https://doc.rust-lang.org/book/ch04-03-slices.html) are not an existing concept in C but could be easily represented as an FFI-safe struct.

> Finally, It’s worth mentioning that there are also proposals to improve the interface between Haskell programs requiring Rust libraries, including this Cabal [RFC](https://github.com/haskell/cabal/issues/7906). As a reminder, the implementation proposed here does not provide support for Haskell dependencies in Rust project yet, but there is a previous [unmaintained attempt](https://github.com/mgattozzi/curryrs) of bringing Haskell runtime support in Rust binaries.

Author: ```suggestion
Finally, It’s worth mentioning that there are also proposals to improve the interface between Haskell programs requiring Rust libraries, including this Cabal [RFC](https://github.com/haskell/cabal/issues/7906). As a reminder, the implementation proposed here does not provide support for Haskell dependencies in Rust project yet, but there is a previous [unmaintained attempt](https://github.com/mgattozzi/curryrs) of bringing Haskell runtime support in Rust binaries. We should also keep a close look on the Rust RFC that offer to introduce a `#[repr(interop)]` attribute: [Experimental feature gate proposal `interoperable_abi`](https://github.com/rust-lang/rust/pull/105586).
```

> 

Author: ```suggestion

I would like to thank [@hsyl20](https://github.com/hsyl20), [@iquerejeta](https://github.com/iquerejeta) and [@govanify](https://github.com/govanify) for their careful reviewing and corrections they provide to make this article in good shape.

```
